libcontainer/configs/config_linux.go

```diff
@@ -22,14 +31,78 @@ type LinuxPersonality struct {
 	Domain int `json:"domain"`
 }
 
+// InsideContainer is set to true if we are running in the context of "runc
+// init". This must only be changed in startInitialization.
+//
+// TODO: Having this live in libcontainer/configs is quite ugly but due to the
+// import graph we can't really move it anywhere else (other than a dedicated
+// package).
+var InsideContainer bool
+
+func mappingFromNamespacePath(nsPath, mapName string) ([]IDMap, error) {
+	var mapData []byte
+	if InsideContainer {
+		// From within the container, we cannot join the target namespace (we
+		// don't have the necessary privileges, and we might be in the
+		// container's mountns after pivot_root(2) so the path might not exist
+		// either).
+		//
+		// If we happen to be inside the namespace in question, we can just
+		// read our own /proc/self/ file. However, calling HostUID and HostGID
+		// from the context of runc init should be avoided because there's no
+		// way for us to be sure that the namespace actually is the same as
+		// nsPath (and isn't different due to a programming bug). So emit a
+		// warning, but return our own mapping since that should still be
+		// correct in all cases.
+		logrus.Infof("cannot robustly retrieve mapping of namespace %s from within runc init, falling back to own mappings (this should work, but is probably a bug in runc or the libcontainer user -- please submit a bug report)", nsPath)
+		data, err := os.ReadFile("/proc/self/" + mapName)
+		if err != nil {
+			return nil, fmt.Errorf("retrieve runc init fallback mapping failed: %w", err)
+		}
+		mapData = data
+	} else {
+		// We have to join the namespace to get the actual mappings.
+		cmd := exec.Command("nsenter", "--no-fork", "--preserve-credentials", "--user="+nsPath, "--", "cat", "/proc/self/"+mapName)
```

Comment:
I think the nsPath is a string like this: /proc/<pid>/ns/user, why not get the pid value from nsPath, and read the content from `/proc/< pid >/(u|g)id_map'? Maybe there is some special conditions I missing?

Reply:
The namespace path can be any path on the system. You can bind-mount namespace files to any other path, and in really extreme cases it might be the case that you don't even have access to the pid in `/proc` that is associated with the namespace (runc runs in a child pidns, and the only process in the target namespace is in a non-child pidns).

The only way to be absolutely sure is to join the namespace.

Reply:
I think in most of time, `nsPath` is a `/proc/< pid >/ns/< ns name >` style, so maybe we should try to read from `/proc/< pid >/< mapName >` first, and then fail back to other solution or just error out. Because:
1. If we don't have access to the pid in /proc, I think `nsenter` also will be fail;
2. The binary `nsenter` and `cat` may not exist in the system.

Reply:
> I think in most of time, nsPath is a /proc/< pid >/ns/< ns name > style, so maybe we should try to read from /proc/< pid >/< mapName > first, and then fail back to other solution

Yeah, that would make this less slow, and is probably a good idea. My only concern is that it means that the `nsenter(1)` path would only be tested in very specific scenarios that we don't have in our integration tests. (I guess I could try to cook up such a test, but running a bunch of subshells in different namespaces is quite complicated to get working in bats.)

I don't think erroring out in this case is okay, because the runtime-spec lets you specify any path.

> If we don't have access to the pid in /proc, I think nsenter also will be fail;

This isn't correct. Being able to see the process in proc doesn't affect whether `setns(2)` works. `nsenter -t` does use the pid (which means it needs to be visible in proc), but if you use `nsenter --user=...` as I do here, then it just uses the path.

> The binary nsenter and cat may not exist in the system.

I implemented this with `nsenter(1)` because it's simpler to get working than the more complicated alternative (make another re-exec target that does this operation a-la `nsexec.c`). It's quite unfortunate that Go doesn't let you spawn containers inside a target namespace...